src/js/map/TerrainGrid.js

```diff
@@ -0,0 +1,140 @@
+import Grid from './Grid'
+
+export default class TerrainGrid extends Grid {
+  constructor(tilesWide, tilesHigh) {
+    super(tilesWide, tilesHigh, 1)
+    this.setTerrainProperties()
+    this.reset()
+  }
+
+  setTerrainProperties() {
+    this.terrainProperties = {
+      normal: {
+        difficulty: 1,
+        color: 0xFFFFFF,
+      },
+      crater: {
+        difficulty: 3,
+        color: 0xCECECE,
+      },
+      ridge: {
+        difficulty: 2,
+        color: 0xE5E5E5,
+      },
+      obstacle: {
+        difficulty: 0,
+        color: 0x000000,
+      },
+    }
+  }
+
+  getTerrainProperties(type) {
+    if (this.terrainProperties.hasOwnProperty(type)) {
+      return {
+        type: type,
+        difficulty: this.terrainProperties[type].difficulty,
+        color: this.terrainProperties[type].color,
+      }
+    }
+    return {
+      type: "normal",
+      difficulty: this.terrainProperties.normal.difficulty,
+      color: this.terrainProperties.normal.color,
+    }
+  }
+
+  reset() {
+    this.initialValue = this.getTerrainProperties("normal")
+    super.reset()
+    this.generateTerrain()
+  }
+
+  generateTerrain() {
+    // TODO: Simple randomizer for terrain features
+    this.addTestCrater({x:17,y:17}, 4)
+    this.addTestCrater({x:30,y:30}, 4)
+    this.addTestCrater({x:30,y:4}, 4)
+  }
+
+  addTestCrater(gridLocation, size) {
+    // square crater
+    for (let x = gridLocation.x - size; x <= gridLocation.x + size; x++) {
+      for (let y = gridLocation.y - size; y <= gridLocation.y + size; y++) {
+        if (!this.coordinateIsValid(x, y)) { continue }
+        if (x == gridLocation.x - size || x == gridLocation.x + size ||
+            y == gridLocation.y - size || y == gridLocation.y + size) {
+          this.set(x, y, this.getTerrainProperties("ridge"))
+        } else if (x == gridLocation.x && y == gridLocation.y) {
+          this.set(x, y, this.getTerrainProperties("obstacle"))
+        } else {
+          this.set(x, y, this.getTerrainProperties("crater"))
+        }
+      }
+    }
+  }
+
+  randomize(wallProbability = 0.1) {
+    for (let i = 0; i < this.tilesWide; i++) {
+      for (let j = 0; j < this.tilesHigh; j++) {
+        let random = Math.random()
+        if (random < wallProbability) {
+          this.set(i, j, this.getTerrainProperties("obstacle"))
+        } else {
+          this.set(i, j, this.getTerrainProperties("normal"))
+        }
+      }
+    }
+  }
+
+  /*
+   * Adds an obstacle onto the weights grid. Simply sets relevant values to 0.
+   * Takes a position, a width, and a height.
+   */
+  addObstacle(gridLocation, gridWidth, gridHeight) {
+    for (let x = gridLocation.x; x < gridLocation.x + gridWidth; x++) {
+      for (let y = gridLocation.y; y < gridLocation.y + gridHeight; y++) {
+        this.set(x, y, this.getTerrainProperties("obstacle"))
+      }
+    }
+  }
+
+  isAreaFree(gridLocation, gridWidth, gridHeight) {
```

Comment:
This method appears to also be in TowerGrid. If we're using it in both, maybe it belongs in Grid?

Reply:
That's true, the main difference is that the version in TerrainGrid specifically checks difficultyAt(), rather than the at() method used in TowerGrid. Just because the two are constructed slightly differently.

Reply:
Fair enough! That's what I get for skimming.